SourceCode/include/tree/KDTree.h

```diff
@@ -0,0 +1,198 @@
+        maxCoordinates[i] = std::max(maxCoordinates[i], vertex->position[i]);
+        minCoordinates[i] = std::min(minCoordinates[i], vertex->position[i]);
+      }
+    }
+    for (auto i = 0; i < 3; i++) {
+      if (minCoordinates[i] > this->maxPoint[i] || maxCoordinates[i] < this->minPoint[i]) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  bool hasIntersection(const Ray &ray) {
+    Vector tMax;  // tXmin, tYmin, tZmin
+    Vector tMin;  // tXmax, tYmax, tZmax
+    for (auto i = 0; i < 3; i++) {
+      if (ray.direction[i] == 0) {
+        if (ray.origin[i] < minPoint[i] || ray.origin[i] > maxPoint[i]) {
+          return false;
+        }
+        tMax[i] = std::numeric_limits<float>::infinity();
+        tMin[i] = -std::numeric_limits<float>::infinity();
+      } else {
+        tMax[i] = (maxPoint[i] - ray.origin[i]) / ray.direction[i];
+        tMin[i] = (minPoint[i] - ray.origin[i]) / ray.direction[i];
+      }
+    }
+
+    for (auto i = 0; i < 3; i++) {
+      if (tMin[i] > 0) {
+        Vector intersectionPoint = ray.origin + tMin[i] * ray.direction;
+        if (intersectionPoint[0] >= this->minPoint[0] && intersectionPoint[0] <= this->maxPoint[0] &&
+            intersectionPoint[1] >= this->minPoint[1] && intersectionPoint[1] <= this->maxPoint[1] &&
+            intersectionPoint[2] >= this->minPoint[2] && intersectionPoint[2] <= this->maxPoint[2]) {
+          return true;
+        }
+      } else if (tMax[i] > 0) {
+        Vector intersectionPoint = ray.origin + tMax[i] * ray.direction;
+        if (intersectionPoint[0] >= this->minPoint[0] && intersectionPoint[0] <= this->maxPoint[0] &&
+            intersectionPoint[1] >= this->minPoint[1] && intersectionPoint[1] <= this->maxPoint[1] &&
+            intersectionPoint[2] >= this->minPoint[2] && intersectionPoint[2] <= this->maxPoint[2]) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+};
+
+class KDTree {
+ private:
+  struct TreeNode {
+    BoundingBox box;
+    int children[2];
+    int parent;
+    std::vector<size_t> triangleIndexes;
+  };
+
+  const Scene &scene;
+  const unsigned short MAX_TRIANGLES_IN_LEAF = 16;
+  const unsigned short MAX_DEPTH = 25;
+  const unsigned short AXIS_COUNT = 3;
+  std::vector<TreeNode> nodes;
+
+  unsigned int createNode(const BoundingBox &box, int firstChildIndex, int secondChildIndex, int parentIndex) {
+    unsigned int newNodeIndex = this->nodes.size();
+    this->nodes.push_back(TreeNode{box, {firstChildIndex, secondChildIndex}, parentIndex, {}});
+    return newNodeIndex;
+  }
+
+  void build(const int parentIndex, const unsigned short depth, const std::vector<size_t> &triangleIndexes) {
+    if (depth >= MAX_DEPTH || triangleIndexes.size() <= MAX_TRIANGLES_IN_LEAF) {
+      this->nodes[parentIndex].triangleIndexes = triangleIndexes;
+      return;
+    }
+
+    std::pair<BoundingBox, BoundingBox> childBoxes = this->nodes[parentIndex].box.split(depth % AXIS_COUNT);
```

Comment:
std::pair could be very unintuitive, because the code quickly gets flooded with "first" and "second", better use simple custom structure.